nova_vm/src/ecmascript/execution/environments/declarative_environment.rs

```diff
@@ -1,56 +1,292 @@
+    /// value V. A binding for N normally already exists, but in rare cases it
+    /// may not. If the binding is an immutable binding, a TypeError is thrown
+    /// if S is true.
+    pub(crate) fn set_mutable_binding(
+        &mut self,
+        agent: &mut Agent,
+        name: Atom,
+        value: Value,
+        mut is_strict: bool,
+    ) -> JsResult<()> {
+        // 1. If envRec does not have a binding for N, then
+        let Some(binding) = self.bindings.get_mut(&name) else {
+            // a. If S is true, throw a ReferenceError exception.
+            if is_strict {
+                return Err(agent
+                    .throw_exception(ExceptionType::ReferenceError, "Identifier is not defined."));
+            }
+
+            // b. Perform ! envRec.CreateMutableBinding(N, true).
+            self.create_mutable_binding(name.clone(), true);
+
+            // c. Perform ! envRec.InitializeBinding(N, V).
+            self.initialize_binding(&name, value);
+
+            // d. Return UNUSED.
+            return Ok(());
+        };
+
+        // 2. If the binding for N in envRec is a strict binding, set S to true.
+        if binding.strict {
+            is_strict = true;
+        }
+
+        // 3. If the binding for N in envRec has not yet been initialized, then
+        if binding.value.is_none() {
+            // a. Throw a ReferenceError exception.
+            return Err(
+                agent.throw_exception(ExceptionType::ReferenceError, "Identifier is not defined.")
+            );
+        }
+
+        // 4. Else if the binding for N in envRec is a mutable binding, then
+        if binding.mutable {
+            // a. Change its bound value to V.
+            binding.value = Some(value);
+        }
+        // 5. Else,
+        else {
+            // a. Assert: This is an attempt to change the value of an immutable binding.
+            debug_assert!(!binding.mutable);
+
+            // b. If S is true, throw a TypeError exception.
+            if is_strict {
+                return Err(
+                    agent.throw_exception(ExceptionType::TypeError, "Cannot assign to constant.")
+                );
+            }
+        }
+
+        // 6. Return UNUSED.
+        Ok(())
+    }
+
+    /// ### [9.1.1.1.6 GetBindingValue ( N, S )](https://tc39.es/ecma262/#sec-declarative-environment-records-getbindingvalue-n-s)
+    ///
+    /// The GetBindingValue concrete method of a Declarative Environment Record
+    /// envRec takes arguments N (a String) and S (a Boolean) and returns either
+    /// a normal completion containing an ECMAScript language value or a throw
+    /// completion. It returns the value of its bound identifier whose name is
+    /// N. If the binding exists but is uninitialized a ReferenceError is
+    /// thrown, regardless of the value of S.
+    pub(crate) fn get_binding_value(
+        &self,
+        agent: &mut Agent,
+        name: &Atom,
+        _is_strict: bool,
```

Comment:
**nitpick:** Since this function doesn't actually use the boolean and we're not implementing these as traits or anything, I'd just drop the parameter: We can leave it as a comment line.